Add wiremock module

mcruzdev opened the pull request:
Many thanks for submitting your Pull Request :heart:!

Please make sure that your PR meets the following requirements:

- [x] You have read the [contributors guide](CONTRIBUTING.md)
- [x] Your code is properly formatted according to [our code style](CONTRIBUTING.md#code-style)
- [x] Pull Request title contains the target branch if not targeting main: `[0.9.x] Subject`
- [x] Pull Request contains link to the issue
- [ ] Pull Request contains link to any dependent or related Pull Request
- [ ] Pull Request contains description of the issue
- [ ] Pull Request does not include fixes for issues other than the main ticket

<details>
<summary>
How to backport a pull request to a different branch?
</summary>

In order to automatically create a **backporting pull request** please add one or more labels having the following format `backport-<branch-name>`, where `<branch-name>` is the name of the branch where the pull request must be backported to (e.g., `backport-quarkus2` to backport the original PR to the `quarkus2` branch).

> **NOTE**: **backporting** is an action aiming to move a change (usually a commit) from a branch (usually the main one) to another one, which is generally referring to a still maintained release branch. Keeping it simple: it is about to move a specific change or a set of them from one branch to another.

Once the original pull request is successfully merged, the automated action will create one backporting pull request per each label (with the previous format) that has been added.

If something goes wrong, the author will be notified and at this point a manual backporting is needed.

> **NOTE**: this automated backporting is triggered whenever a pull request on `main` branch is labeled or closed, but both conditions must be satisfied to get the new PR created.
</details>

mcruzdev commented:
Hi @hbelmiro and @ricardozanini I created the wiremock module, what initial features we need to have to merge to main?

Is it possible to create a branch for wiremock stuff, or just adding an initial feature is ok for now?

hbelmiro commented:
@mcruzdev 

> what initial features we need to have to merge to main?

From us, you're probably the one who knows more about it. I'll let you decide.
Is this wiremock support a real need for you as a user? If so, what are the ones you need the most? From those, pick the easiest/fastest to implement.

> Is it possible to create a branch for wiremock stuff, or just adding an initial feature is ok for now?
IMO, we can add features incrementally (an initial feature is ok), so users can benefit from them ASAP.

mcruzdev commented:
Hi @hbelmiro, Thank you!

It help me a lot, I will continue here!

mcruzdev commented:
- [ ] Add response status from OpenAPI
- [ ] Add response body from OpenAPI (without to get from components)

mcruzdev commented:
Hi @hbelmiro and @ricardozanini, I think that we can get a first review about this new feature.

ricardozanini commented:
I'd rather wait for #482 to be merged first. @carlesarnal will kick my ass if I keep asking him to rebase.

mcruzdev commented on wiremock/deployment/pom.xml:
```diff
@@ -0,0 +1,61 @@
+    <name>Quarkus - Openapi Generator - Wiremock - Deployment</name>
+
+    <properties>
+        <version.io.swagger.parser>2.1.20</version.io.swagger.parser>
```
This is the same used by `client`.

carlesarnal replied:
No, we can't replace this.

mcruzdev commented:
> Just a few comments initially, I'll do a better review later.

I switched the extension to `experimental` Sound good?

I've added some basic features for starters. It is great for new contributors!

ricardozanini commented:
@mcruzdev can you rebase? 🙏

ricardozanini commented:
@mcruzdev are the issues pointed by GH relevant?

mcruzdev commented:
> @mcruzdev are the issues pointed by GH relevant?

I think we can ignore because are Test classes. makes sense?

> Unused class: SchemaReaderTest is not referenced within this codebase. If not used as an external API it should be removed.

hbelmiro commented on wiremock/deployment/src/main/java/io/quarkiverse/openapi/wiremock/generator/deployment/wiremock/model/Request.java:
```diff
@@ -0,0 +1,40 @@
+
+import java.util.Objects;
+
+public class Request {
```
Can't this be a `record`?

hbelmiro commented on wiremock/deployment/src/main/java/io/quarkiverse/openapi/wiremock/generator/deployment/wiremock/model/Response.java:
```diff
@@ -0,0 +1,28 @@
+import com.fasterxml.jackson.annotation.JsonInclude;
+
+@JsonInclude(JsonInclude.Include.NON_EMPTY)
+public class Response {
```
Can't this be a `record`?

mcruzdev replied:
Hi @hbelmiro, yeah can be. Another point here is the use of `@JsonInclude` annotation, I do no know if makes sense to use it, I will try to remove it and use another way to not include non empty.

hbelmiro commented on wiremock/deployment/src/main/java/io/quarkiverse/openapi/wiremock/generator/deployment/wiremock/model/Stubbing.java:
```diff
@@ -0,0 +1,43 @@
+ * See <a href="https://wiremock.org/docs/stubbing/#basic-stubbing">here</a>
+ * the official Wiremock documentation for more details.
+ */
+public record Stubbing(Request request, Response response) {
```
I think we should rename it to `Stub`.

hbelmiro commented on wiremock/deployment/src/main/java/io/quarkiverse/openapi/wiremock/generator/deployment/wiremock/model/Stubbing.java:
```diff
@@ -0,0 +1,43 @@
+        return this.request.getMethod();
+    }
+
+    public static class StubbingBuilder {
```
Do we really need a builder?

mcruzdev replied:
Actually no, we can add it when we have more attribute here.

hbelmiro commented on wiremock/deployment/src/main/java/io/quarkiverse/openapi/wiremock/generator/deployment/wrapper/OpenApiWiremockGeneratorWrapper.java:
```diff
@@ -0,0 +1,91 @@
+    private static final Logger LOGGER = LoggerFactory.getLogger(OpenApiWiremockGeneratorWrapper.class);
+    public static final String MAPPINGS_DIR = "mappings";
+    private static final String MAPPINGS = "mappings";
+    private static final String WIREMOCK_OUTPUT_FILE = "wiremock-stubbings.json";
```
```suggestion
    private static final String WIREMOCK_OUTPUT_FILE = "wiremock-stubs.json";
```

hbelmiro commented on wiremock/deployment/src/main/java/io/quarkiverse/openapi/wiremock/generator/deployment/wrapper/OpenApiWiremockGeneratorWrapper.java:
```diff
@@ -0,0 +1,91 @@
+                Files.write(wiremockStubbings, json);
+
+            } catch (IOException e) {
+                LOGGER.warn("Failed to generate Wiremock stubs for file " + file.toAbsolutePath(), e);
```
Are you sure we don't need to abort the entire generation here?

mcruzdev replied:
We are generating one Wiremock stub for one OpenAPI specification file, the generation is independent, if one fails does not affect the another one, in your experience, is better make the generation dependent?

hbelmiro replied:
In general, I prefer failing "atomically".
In this particular case, it doesn't seem like users would need to generate just a few stubs. I think they'd like to generate them all. (I would, as a user. But I may be wrong.)
Also, if we just log the exception, users may not notice it occurred.
I don't have a strong opinion here, but I would generate them all or fail (not just log the error).

mcruzdev replied:
It help me to think, makes sense!

hbelmiro commented on wiremock/docs/antora.yaml:
```diff
@@ -0,0 +1,5 @@
+name: quarkus-openapi-generator
+title: Openapi Generator
```
```suggestion
title: OpenAPI Generator
```

hbelmiro commented on wiremock/docs/modules/ROOT/nav.adoc:
```diff
@@ -0,0 +1 @@
+* xref:index.adoc[Quarkus - Openapi Generator - Wiremock]
```
```suggestion
* xref:index.adoc[Quarkus - OpenAPI Generator - Wiremock]
```

hbelmiro commented on wiremock/docs/modules/ROOT/pages/index.adoc:
```diff
@@ -0,0 +1,22 @@
+= Quarkus - Openapi Generator - Wiremock
```
```suggestion
= Quarkus - OpenAPI Generator - Wiremock
```

hbelmiro commented on wiremock/integration-tests/pom.xml:
```diff
@@ -0,0 +1,95 @@
+    <modelVersion>4.0.0</modelVersion>
+
+    <artifactId>quarkus-openapi-generator-wiremock-integration-tests</artifactId>
+    <name>Quarkus - Openapi Generator - Wiremock - Integration Tests</name>
```
```suggestion
    <name>Quarkus - OpenAPI Generator - Wiremock - Integration Tests</name>
```

hbelmiro commented on wiremock/pom.xml:
```diff
@@ -0,0 +1,22 @@
+    <modelVersion>4.0.0</modelVersion>
+
+    <artifactId>quarkus-openapi-generator-wiremock-parent</artifactId>
+    <name>Quarkus - Openapi Generator - Wiremock - Parent</name>
```
```suggestion
    <name>Quarkus - OpenAPI Generator - Wiremock - Parent</name>
```

hbelmiro commented on wiremock/runtime/pom.xml:
```diff
@@ -0,0 +1,132 @@
+        <relativePath>../pom.xml</relativePath>
+    </parent>
+    <artifactId>quarkus-openapi-generator-wiremock</artifactId>
+    <name>Quarkus - Openapi Generator - Wiremock</name>
```
```suggestion
    <name>Quarkus - OpenAPI Generator - Wiremock</name>
```

hbelmiro reviewed (commented):
Good job and amazing contribution @mcruzdev. Thank you!
I left some comments.

mcruzdev commented:
Hi @hbelmiro thank you for the reviewing it, I will take a look soon!

mcruzdev commented on wiremock/deployment/src/main/java/io/quarkiverse/openapi/wiremock/generator/deployment/wiremock/OpenApi2WiremockMapper.java:
```diff
@@ -0,0 +1,164 @@
+        return stubbings.stream().toList();
+    }
+
+    private String getResponseBodyFrom$Ref(String component) {
```
No Helber, I wanted to follow the OpenAPI specification name, I can change here.

hbelmiro replied:
The compiler uses `$` to name inner classes. I'm not sure it would cause issues in real life, but I'd avoid it.
It may also confuse log readers as well. They may think that `Ref` is an inner class.

mcruzdev replied:
It makes total sense!

hbelmiro commented:
@mcruzdev can you please let me know when I can re-review it?

mcruzdev commented on wiremock/deployment/src/main/java/io/quarkiverse/openapi/wiremock/generator/deployment/wiremock/SchemaReader.java:
```diff
@@ -0,0 +1,39 @@
+import io.swagger.v3.oas.models.media.Schema;
+
+public class SchemaReader {
+
```
I changed this class to be an interface instead

mcruzdev commented on wiremock/deployment/src/main/java/io/quarkiverse/openapi/wiremock/generator/deployment/wiremock/SchemaReader.java:
```diff
@@ -0,0 +1,39 @@
+
+    public static String readObjectExample(Schema<?> schema) {
+        try {
+            HashMap<String, Object> map = new HashMap<>();
```
I did not understand this point @hbelmiro, can you explain in more details?

hbelmiro replied:
We can remove the `Map` parameter of `mapObjectExample`.
By doing this, we avoid modifying objects passed as an argument, increasing the code's immutability.
Like this:

```java
    static String readObjectExample(Schema<?> schema) {
        try {
            Map<String, Object> map = mapObjectExample(schema);
            return OpenApiWiremockGeneratorWrapper.OBJECT_MAPPER_INSTANCE.writeValueAsString(map);
        } catch (JsonProcessingException e) {
            return EMPTY_JSON_OBJECT;
        }
    }

    private static Map<String, Object> mapObjectExample(Schema<?> example) {
        HashMap<String, Object> root = new HashMap<>();
        Optional.ofNullable(example.getProperties()).orElse(Map.of())
                .forEach((key, schema) -> {
                    if (schema.getType().equals(OBJECT_TYPE)) {
                        root.put(key, mapObjectExample(schema));
                    } else {
                        root.put(key, schema.getExample());
                    }
                });
        return root;
    }
```

hbelmiro commented on wiremock/deployment/src/main/java/io/quarkiverse/openapi/wiremock/generator/deployment/wiremock/OpenApi2WiremockMapper.java:
```diff
@@ -0,0 +1,160 @@
+package io.quarkiverse.openapi.wiremock.generator.deployment.wiremock;
+
+import static io.swagger.v3.parser.util.SchemaTypeUtil.*;
```
Please don't use star imports.

hbelmiro commented on wiremock/deployment/src/main/java/io/quarkiverse/openapi/wiremock/generator/deployment/wiremock/model/Request.java:
```diff
@@ -0,0 +1,15 @@
+        this.method = Objects.requireNonNull(method);
+    }
+
+    public static Request create(final String urlPath, final String method) {
```
It seems like this can be removed.

mcruzdev replied:
>We can remove the Map parameter of mapObjectExample.
By doing this, we avoid modifying objects passed as an argument, increasing the code's immutability.
Like this:

It makes total sense, TY for this.

mcruzdev commented:
Hi @hbelmiro and @ricardozanini I think we can wait for merge this pull request, I am seeing here some projects that does similar things and I am collecting some ideas to use here... Another point is: I am very busy and I think if we merge it we will not have effort to upgrade this module faster. WDYT?
